tests/tests/classes/PluginManager.js

```diff
@@ -0,0 +1,281 @@
+
+      const servicePluginMock1 = new ServicePluginMock1();
+      const servicePluginMock2 = new ServicePluginMock2();
+
+      expect(pluginManager.plugins).to.contain(servicePluginMock1);
+      expect(pluginManager.plugins).to.contain(servicePluginMock2);
+      expect(pluginManager.plugins).to.contain(helloWorld);
+    });
+  });
+
+  describe('#loadCorePlugins()', () => {
+    it('should load the Serverless core plugins', () => {
+      pluginManager.loadCorePlugins();
+
+      expect(pluginManager.plugins).to.contain(helloWorld);
+    });
+  });
+
+  describe('#loadServicePlugins()', () => {
+    it('should load the service plugins', () => {
+      const servicePlugins = [ServicePluginMock1, ServicePluginMock2];
+      pluginManager.loadServicePlugins(servicePlugins);
+
+      const servicePluginMock1 = new ServicePluginMock1();
+      const servicePluginMock2 = new ServicePluginMock2();
+
+      expect(pluginManager.plugins).to.contain(servicePluginMock1);
+      expect(pluginManager.plugins).to.contain(servicePluginMock2);
+    });
+  });
+
+  describe('#loadCommands()', () => {
+    it('should load the plugin commands', () => {
+      pluginManager.loadCommands(SynchronousPluginMock);
+
+      expect(pluginManager.commandsList[0]).to.have.property('deploy');
+    });
+  });
+
+  describe('#getEvents()', () => {
+    beforeEach(() => {
+      pluginManager.loadCommands(SynchronousPluginMock);
+    });
+
+    it('should get all the matching events for a root level command', () => {
+      const commandsArray = ['deploy'];
+      const events = pluginManager.getEvents(commandsArray, pluginManager.commands);
+
+      expect(events).to.contain('before:deploy:resources');
+      expect(events).to.contain('deploy:resources');
+      expect(events).to.contain('after:deploy:resources');
+      expect(events).to.contain('before:deploy:functions');
+      expect(events).to.contain('deploy:functions');
+      expect(events).to.contain('after:deploy:functions');
+    });
+
+    it('should get all the matching events for a nested level command', () => {
+      const commandsArray = ['deploy', 'onpremises'];
+      const events = pluginManager.getEvents(commandsArray, pluginManager.commands);
+
+      expect(events).to.contain('before:deploy:onpremises:resources');
+      expect(events).to.contain('deploy:onpremises:resources');
+      expect(events).to.contain('after:deploy:onpremises:resources');
+      expect(events).to.contain('before:deploy:onpremises:functions');
+      expect(events).to.contain('deploy:onpremises:functions');
+      expect(events).to.contain('after:deploy:onpremises:functions');
+    });
+
+    it('should return an empty events array when the command is not defined', () => {
+      const commandsArray = ['foo'];
+      const events = pluginManager.getEvents(commandsArray, pluginManager.commands);
+
+      expect(events.length).to.equal(0);
+    });
+  });
+
+  describe('#runCommand()', () => {
```

Comment:
For the hooks its also important that they hooks are called in the same order that we've added the plugins to the system. This is really key to make sure that there aren't any unexpected changes to the order plugins are run for users